thread_sync.c

```diff
@@ -839,15 +839,29 @@ queue_closed_result(VALUE self, struct rb_queue *q)
 /*
  * Document-method: Queue::new
  *
- * Creates a new queue instance.
+ * Creates a new queue instance, optionally using the contents of an Enumerable
+ * for its initial state.
+ *
+ *  Example:
+ *
+ *    	q = Queue.new
+ *    	q = Queue.new([a, b, c])
+ *    	q = Queue.new(items)
  */
 
 static VALUE
-rb_queue_initialize(VALUE self)
+rb_queue_initialize(int argc, VALUE *argv, VALUE self)
 {
+    VALUE initial;
     struct rb_queue *q = queue_ptr(self);
     RB_OBJ_WRITE(self, &q->que, ary_buf_new());
     list_head_init(queue_waitq(q));
+    rb_scan_args(argc, argv, "01", &initial);
+    if (argc == 1) {
+        RB_OBJ_WRITE(self, &q->que, rb_convert_type(initial, T_ARRAY, "Array", "to_ary"));
+    } else {
+        RB_OBJ_WRITE(self, &q->que, ary_buf_new());
```

Comment:
else block seems unnecessary

Reply:
Gah sorry - actually it's the previous assignment of `q->que` that was redundant. I've pushed. What we're doing is setting the queue's array in the first place to a `dup` of the array from the `Enumerable`, or a new empty array.

Reply:
`ary_buf_new` creates a hidden array.
I think concatenation to it is safer than dup.

Reply:
Thanks done.